File: samples/grpc/restaurant-drone-deliveries-service-scala/src/main/scala/central/drones/Drone.scala
package central.drones

import java.time.Instant

import scala.concurrent.duration._

import akka.Done
import akka.actor.typed.SupervisorStrategy
import akka.actor.typed.scaladsl.ActorContext
import akka.actor.typed.scaladsl.Behaviors
import akka.actor.typed.ActorRef
import akka.actor.typed.ActorSystem
import akka.actor.typed.Behavior
import akka.cluster.sharding.typed.scaladsl.ClusterSharding
import akka.cluster.sharding.typed.scaladsl.Entity
import akka.cluster.sharding.typed.scaladsl.EntityTypeKey
import akka.pattern.StatusReply
import akka.persistence.r2dbc.state.scaladsl.AdditionalColumn
import akka.persistence.typed.PersistenceId
import akka.persistence.typed.state.scaladsl.DurableStateBehavior
import akka.persistence.typed.state.scaladsl.Effect
import akka.serialization.jackson.CborSerializable
import central.CoarseGrainedCoordinates

/**
 * Durable state entity keeping an overview state of where the drone is and its state, but not the full detail,
 * that stays in the local control service.
 */
object Drone {

  // #commands
  sealed trait Command extends CborSerializable

  final case class UpdateLocation(
      locationName: String,
      coarseGrainedCoordinates: CoarseGrainedCoordinates,
      replyTo: ActorRef[StatusReply[Done]])
      extends Command

  final case class GetState(replyTo: ActorRef[State]) extends Command
  // #commands

  val EntityKey = EntityTypeKey[Command]("CentralDrone")

  // #state
  final case class State(
      locationName: String,
      currentLocation: Option[CoarseGrainedCoordinates],
      lastChange: Instant)
      extends CborSerializable
  // #state

  // #emptyState
  private val emptyState = State("unknown", None, Instant.EPOCH)
  // #emptyState

  def init(system: ActorSystem[_]): Unit = {
    ClusterSharding(system).init(Entity(EntityKey)(entityContext =>
      Drone(entityContext.entityId)))
  }

  def apply(droneId: String): Behavior[Command] = {
    Behaviors.setup { context =>
      DurableStateBehavior(
        PersistenceId(EntityKey.name, droneId),
        emptyState,
        onCommand(context)).onPersistFailure(
        SupervisorStrategy.restartWithBackoff(100.millis, 5.seconds, 0.1))
    }
  }

  // #commandHandler
  private def onCommand(context: ActorContext[Command])(
      state: State,
      command: Command): Effect[State] =
    command match {
      case UpdateLocation(locationName, coordinates, replyTo) =>
        context.log.info(
          "Updating location to [{}], [{}]",
          locationName,
          coordinates)
        Effect
          .persist(
            state.copy(
              locationName = locationName,
              currentLocation = Some(coordinates)))
          .thenReply(replyTo)(_ => StatusReply.ack())

      case GetState(replyTo) =>
        Effect.reply(replyTo)(state)
    }
  // #commandHandler

}

// #locationColumn
/**
 * Write local drone control location name column for querying drone locations per control location
 */
final class LocationColumn extends AdditionalColumn[Drone.State, String] {

  override def columnName: String = "location"

  override def bind(upsert: AdditionalColumn.Upsert[Drone.State])
      : AdditionalColumn.Binding[String] =
    AdditionalColumn.BindValue(upsert.value.locationName)

}
// #locationColumn

File: samples/grpc/restaurant-drone-deliveries-service-java/src/main/java/central/drones/Drone.java
package central.drones;

import akka.Done;
import akka.actor.typed.ActorRef;
import akka.actor.typed.ActorSystem;
import akka.actor.typed.SupervisorStrategy;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.cluster.sharding.typed.javadsl.ClusterSharding;
import akka.cluster.sharding.typed.javadsl.Entity;
import akka.cluster.sharding.typed.javadsl.EntityTypeKey;
import akka.pattern.StatusReply;
import akka.persistence.typed.PersistenceId;
import akka.persistence.typed.state.javadsl.CommandHandler;
import akka.persistence.typed.state.javadsl.DurableStateBehavior;
import akka.persistence.typed.state.javadsl.Effect;
import akka.serialization.jackson.CborSerializable;
import central.CoarseGrainedCoordinates;
import com.fasterxml.jackson.annotation.JsonCreator;
import java.time.Duration;
import java.time.Instant;
import java.util.Optional;

public final class Drone extends DurableStateBehavior<Drone.Command, Drone.State> {

  // #commands
  interface Command extends CborSerializable {}

  public static final class UpdateLocation implements Command {
    public final String locationName;
    public final CoarseGrainedCoordinates coarseGrainedCoordinates;
    public final ActorRef<StatusReply<Done>> replyTo;

    public UpdateLocation(
        String locationName,
        CoarseGrainedCoordinates coarseGrainedCoordinates,
        ActorRef<StatusReply<Done>> replyTo) {
      this.locationName = locationName;
      this.coarseGrainedCoordinates = coarseGrainedCoordinates;
      this.replyTo = replyTo;
    }
  }

  public static final class GetState implements Command {
    public final ActorRef<State> replyTo;

    @JsonCreator
    public GetState(ActorRef<State> replyTo) {
      this.replyTo = replyTo;
    }
  }

  // #commands

  // #state
  public static final class State implements CborSerializable {
    public String locationName;
    public Optional<CoarseGrainedCoordinates> currentLocation;
    public Instant lastChange;

    public State(
        String locationName,
        Optional<CoarseGrainedCoordinates> currentLocation,
        Instant lastChange) {
      this.locationName = locationName;
      this.currentLocation = currentLocation;
      this.lastChange = lastChange;
    }
  }

  // #state

  public static final EntityTypeKey<Command> ENTITY_KEY =
      EntityTypeKey.create(Command.class, "CentralDrone");

  public static void init(ActorSystem<?> system) {
    ClusterSharding.get(system)
        .init(
            Entity.of(
                ENTITY_KEY,
                entityContext ->
                    Behaviors.setup(context -> new Drone(context, entityContext.getEntityId()))));
  }

  private final ActorContext<Command> context;

  private Drone(ActorContext<Command> context, String entityId) {
    super(
        PersistenceId.of(ENTITY_KEY.name(), entityId),
        SupervisorStrategy.restartWithBackoff(Duration.ofMillis(100), Duration.ofSeconds(5), 0.1));
    this.context = context;
  }

  // #emptyState
  @Override
  public State emptyState() {
    return new State("unknown", Optional.empty(), Instant.EPOCH);
  }

  // #emptyState

  // #commandHandler
  @Override
  public CommandHandler<Command, State> commandHandler() {
    return newCommandHandlerBuilder()
        .forAnyState()
        .onCommand(UpdateLocation.class, this::onUpdateLocation)
        .onCommand(
            GetState.class,
            (state, command) ->
                // reply with defensive copy since state is mutable
                Effect()
                    .reply(
                        command.replyTo,
                        new State(state.locationName, state.currentLocation, state.lastChange)))
        .build();
  }

  private Effect<State> onUpdateLocation(State state, UpdateLocation command) {
    context
        .getLog()
        .info(
            "Updating location to [{}], [{}]",
            command.locationName,
            command.coarseGrainedCoordinates);
    state.locationName = command.locationName;
    state.currentLocation = Optional.of(command.coarseGrainedCoordinates);
    state.lastChange = Instant.now();
    return Effect().persist(state).thenReply(command.replyTo, updatedState -> StatusReply.ack());
  }
  // #commandHandler
}
